### PR TITLE
Refactor OperatingSystemConfig deployment into component

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area dev-productivity quality robustness
/kind technical-debt test
/priority normal
/topology seed
/platform all
/exp expert
/merge squash

**What this PR does / why we need it**:
This PR refactors the deployment for the `OperatingSystemConfig` extension resource and adopts the component concept, similar to #3222 and others. This improves several aspects and the overall code quality.

Unlike to other refactoring PRs for extension resources, the  preconditions were different because the `OperatingSystemConfig` was applied as part of the `seed-operatingsystemconfig` chart. Hence, I had to refactor this in its entirety and decided to eliminate the whole chart, eventually. As a result, the PR is quite large, however, we now have unit tests for the vast majority of this part of the the code.

**Which issue(s) this PR fixes**:
Part of #2754

**Special notes for your reviewer**:
- ❗ Due to the large size of the PR I've split it into a bunch of separate, self-contained commits to make the review process very easy. Consider reviewing the various commits individually.
- The `OperatingSystemConfig` resource was the last extension resource that was required to refactor into a component. This allows us to get rid of some control plane migration code and to simply the respective function invocations.
- I'm introducing the `github.com/go-bindata/go-bindata` as new dev tool. With Golang 1.16 we can probably get rid of it in favor of `//go:embed`.
- I have moved some functions to hopefully better places (e.g., from the extensions library to the new `operatingsystemconfig` package) to improve the package dependency graph.
- I have slightly refactored a few of the scripts that we apply in the `original` `OperatingSystemConfig` because we are now injecting Golang constants into them, hence, we don't need Bash variables for certain paths anymore.
- I have split up the `health-monitor` script which was used for `docker`, `containerd`, and `kubelet` into three dedicated scripts and added them to respective component packages.
- The `operatingsystemconfig/original/components/logrotate` component does not implement the new `operatingsystemconfig/original/components.Component` interface because it is just a "helper component" for both `docker` and `containerd`.
- I am aware of a few inconsistencies in the handling/implementation of the various components for extension resources and I plan to address them in future PRs.

**Release note**:

```breaking dependency
The `FileContentInlineCodec` and `UnitSerializer` interfaces have been moved from `extensions/pkg/webhook/controlplane` to `pkg/operation/botanist/extensions/operatingsystemconfig/utils`.
```
```breaking dependency
The `KubeletConfigCodec` interface has been moved from `extensions/pkg/webhook/controlplane` to `pkg/operation/botanist/extensions/operatingsystemconfig/original/components/kubelet` and was renamed to `ConfigCodec`.
```
```breaking developer
The `CopyValues` function has been moved from `pkg/utils/chart` to `pkg/utils` and was renamed to `ShallowCopyMapStringInterface`.
```
```breaking developer
The `ImageMapToValues` function has been moved from `pkg/utils/chart` to `pkg/utils/imagevector`.
```
```feature developer
The `pkg/utils/kubernetes` does now have a new `DeleteObjectsFromListConditionally` utility function.
```

### CODE REVIEW

**Comment by rfranzke** [comment]
After the merge of #3449 I've now rebased this PR and it's ready for review.
/cc @vpnachev

File: pkg/gardenlet/controller/shoot/shoot_control_reconcile.go
Comment (vpnachev) [inline pkg/gardenlet/controller/shoot/shoot_control_reconcile.go]:
```suggestion
			Name:         "Deployibg operating system specific configuration for shoot workers",
```

**Review by vpnachev — DISMISSED** [review]
/lgtm

I have one change request, but if you want, it can be contributed with some other PR.

Personally, I would prefer to keep the commits - this was very helpful for me to go through the changes. If we need to re-check later what was changed, it will be again nice to have everything split in the dedicated commits.

**Comment by rfranzke** [comment]
Thanks for your review @vpnachev!

> I have one change request, but if you want, it can be contributed with some other PR.

Sure, let's wait if there are other comments and if yes then I'll combine it when addressing them.

> Personally, I would prefer to keep the commits - this was very helpful for me to go through the changes. If we need to re-check later what was changed, it will be again nice to have everything split in the dedicated commits.

Sure, OK for me, no strong opinion. If you prefer to keep them then let's do it. @gardener/gardener-maintainers anybody else planning a review and having a preference here as well?

File: pkg/operation/botanist/extensions/operatingsystemconfig/original/components/docker/component.go
Comment (ialidzhikov) [inline pkg/operation/botanist/extensions/operatingsystemconfig/original/components/docker/component.go]: You could move this to the init func below. No need to make it a pkg var.

```suggestion
```

File: pkg/operation/botanist/extensions/operatingsystemconfig/original/components/containerd/component.go
Comment (ialidzhikov) [inline pkg/operation/botanist/extensions/operatingsystemconfig/original/components/containerd/component.go]: You could move below to the init func.

```suggestion
```

File: pkg/operation/botanist/extensions/operatingsystemconfig/original/components/kubelet.go
Comment (ialidzhikov) [inline pkg/operation/botanist/extensions/operatingsystemconfig/original/components/kubelet.go]:
```suggestion
// KubeletConfigParametersFromCoreV1beta1KubeletConfig computes the ConfigurableKubeletConfigParameters based on the provided
```

File: pkg/operation/botanist/extensions/operatingsystemconfig/original/components/kubelet/component.go
Comment (ialidzhikov) [inline pkg/operation/botanist/extensions/operatingsystemconfig/original/components/kubelet/component.go]:
```suggestion
```

File: pkg/operation/botanist/extensions/operatingsystemconfig/downloader/downloader.go
Comment (ialidzhikov) [inline pkg/operation/botanist/extensions/operatingsystemconfig/downloader/downloader.go]:
```suggestion
```

File: pkg/operation/botanist/extensions/operatingsystemconfig/executor/executor.go
Comment (ialidzhikov) [inline pkg/operation/botanist/extensions/operatingsystemconfig/executor/executor.go]:
```suggestion
```

File: pkg/apis/extensions/v1alpha1/types_operatingsystemconfig.go
Comment (ialidzhikov) [inline pkg/apis/extensions/v1alpha1/types_operatingsystemconfig.go:226]: `test/integration/shoots/operations/containerruntime.go` needs to be adapted as it is using `CRINameContainerD` to assert the Node label `worker.gardener.cloud/cri-name` value.

```
• Failure [277.286 seconds]
Shoot container runtime testing
/go/src/github.com/gardener/gardener/test/integration/shoots/operations/containerruntime.go:30
  [DEFAULT] [SERIAL] [SHOOT] should add worker pool with containerd [It]
  /go/src/github.com/gardener/gardener/test/framework/gingko_utils.go:26

  Expected
      <string>: containerd
  to equal
      <v1alpha1.CRIName>: containerd

  /go/src/github.com/gardener/gardener/test/integration/shoots/operations/containerruntime.go:98
```

**Review by ialidzhikov — CHANGES_REQUESTED** [review]
Looks good to me in general. Minor nits inline.

**Review by vpnachev — DISMISSED** [review]
/lgtm

**Review by ialidzhikov — DISMISSED** [review]
/lgtm

**Comment by rfranzke** [comment]
Thanks for approving. I guess, we can merge the PR then. @ialidzhikov (and others, if interested), do you also have an opinion about squash vs. keep-commits?

**Comment by ialidzhikov** [comment]
> Thanks for approving. I guess, we can merge the PR then. @ialidzhikov (and others, if interested), do you also have an opinion about squash vs. keep-commits?

I would also vote for keeping the commits as it was also very helpful for me to go through separate commits while checking this PR.
A rebase is needed meanwhile.

**Comment by rfranzke** [comment]
OK, then let's do it like this. I have rebased and squashed the "address PR review feedback" commits into the original commits now.

/remove merge/squash
/add merge/keep-commits

**Review by vpnachev — DISMISSED** [review]
/lgtm

File: pkg/operation/botanist/extensions/operatingsystemconfig/operatingsystemconfig.go
Comment (danielfoehrKn) [inline pkg/operation/botanist/extensions/operatingsystemconfig/operatingsystemconfig.go:418]:
```suggestion
		globalKubeletConfigParameters: kubeletConfigParameters,
```

Makes it a bit more clear?

File: pkg/operation/botanist/extensions/operatingsystemconfig/operatingsystemconfig.go
Comment (danielfoehrKn) [inline pkg/operation/botanist/extensions/operatingsystemconfig/operatingsystemconfig.go:457]:
```suggestion
	seedClient    client.Client
```

File: pkg/operation/botanist/extensions/operatingsystemconfig/operatingsystemconfig.go
Comment (danielfoehrKn) [inline pkg/operation/botanist/extensions/operatingsystemconfig/operatingsystemconfig.go:460]: Can we adjust the naming or add a comment to make more clear what this "key" is? 
This is the unique key based on the worker name and K8s version and 
 - is the name of the secret containing the cloud-config-execution script for a worker pool
 - serves a  unique prefix for the OSC resource name which consists of: <key>-<original/downloader>

File: pkg/operation/botanist/extensions/operatingsystemconfig/original/components/kubelet/cliflags.go
Comment (danielfoehrKn) [inline pkg/operation/botanist/extensions/operatingsystemconfig/original/components/kubelet/cliflags.go]: Have socket as a constant? Currently we do not allow to change it.

File: pkg/operation/botanist/extensions/operatingsystemconfig/original/components/components.go
Comment (danielfoehrKn) [inline pkg/operation/botanist/extensions/operatingsystemconfig/original/components/components.go:27]: Does not seem to be used anywhere, can we remove it then?

**Review by danielfoehrKn — DISMISSED** [review]
Really nice refactoring. 

Just some minor remarks. Will test it tomorrow morning if that is ok, I know the PR has been open for quite a while.

File: pkg/operation/botanist/extensions/operatingsystemconfig/original/components/components.go
Comment (rfranzke) [reply pkg/operation/botanist/extensions/operatingsystemconfig/original/components/components.go:27]: I used it here https://github.com/rfranzke/gardener/commit/740d911b34ea0794208f056e38f48c74500d03b3#diff-c8b75044fad4e522e6c5fff61acba3554798a4c78e4a649a779b5520d72dd9aaR108-R143 to test the correct order of the components. Yes, the `Name() string` function is not used in the "production" code, however, it's very lightweight, so I guess it's fine. If you have better proposals for the test, feel free to shout it out.

File: pkg/operation/botanist/extensions/operatingsystemconfig/operatingsystemconfig.go
Comment (rfranzke) [reply pkg/operation/botanist/extensions/operatingsystemconfig/operatingsystemconfig.go:460]: Hm, I don't really want to add such comment here. It is used in various places and there is already a comment here https://github.com/rfranzke/gardener/commit/a28511722bf04a9803978834894e11fa60682dcb#diff-a4b93797c6ef59074974f40fbc4b8b7d15661a91b978bcce5637ebfd3de83e38R593.

File: pkg/operation/botanist/extensions/operatingsystemconfig/operatingsystemconfig.go
Comment (rfranzke) [reply pkg/operation/botanist/extensions/operatingsystemconfig/operatingsystemconfig.go:457]: Hm, no, it's called `client` everywhere (also in other components, e.g. https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/extensions/controlplane/controlplane.go#L91), so let's keep it like this.

File: pkg/operation/botanist/extensions/operatingsystemconfig/operatingsystemconfig.go
Comment (rfranzke) [reply pkg/operation/botanist/extensions/operatingsystemconfig/operatingsystemconfig.go:418]: Not sure, I would like to keep it similar to the `kubeletCLIFlags` field.

File: pkg/operation/botanist/extensions/operatingsystemconfig/original/components/kubelet/cliflags.go
Comment (rfranzke) [reply pkg/operation/botanist/extensions/operatingsystemconfig/original/components/kubelet/cliflags.go]: Yes, makes sense, done.

**Comment by rfranzke** [comment]
> Just some minor remarks. Will test it tomorrow morning if that is ok, I know the PR has been open for quite a while.

Well, indeed. It actually was already approved earlier and ready to merge.

**Review by vpnachev — APPROVED** [review]
/lgtm

**Pull request merged**
